In [ ]:
import pandas as pd

# Load the dataset
file_path = '/content/Auto.csv'
auto_df = pd.read_csv(file_path)

# Display the first few rows to understand the structure of the data
auto_df.head()

mpg  cylinders  displacement horsepower  weight  acceleration  year  \
0  18.0          8         307.0        130    3504          12.0    70   
1  15.0          8         350.0        165    3693          11.5    70   
2  18.0          8         318.0        150    3436          11.0    70   
3  16.0          8         304.0        150    3433          12.0    70   
4  17.0          8         302.0        140    3449          10.5    70   

   origin                       name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  
3       1              amc rebel sst  
4       1                ford torino

In [ ]:
from sklearn.preprocessing import StandardScaler

# Identify features for scaling (excluding non-numeric and target variables)
features_to_scale = ['cylinders', 'displacement', 'weight', 'acceleration', 'year']
# Exclude 'mpg' since it will be used for the binary variable creation

# Scale features to mean 0 and standard deviation 1
scaler = StandardScaler()
auto_df[features_to_scale] = scaler.fit_transform(auto_df[features_to_scale])

# Create binary variable for 'mpg' above/below median
mpg_median = auto_df['mpg'].median()
auto_df['high_mpg'] = (auto_df['mpg'] > mpg_median).astype(int)

# Display the updated DataFrame to the user
print("Auto Dataset with Scaled Features and Binary MPG:") # Added a descriptive message
display(auto_df)
print(auto_df)


Auto Dataset with Scaled Features and Binary MPG:


mpg  cylinders  displacement horsepower    weight  acceleration  \
0    18.0   1.495535      1.088435        130  0.630274     -1.294604   
1    15.0   1.495535      1.500913        165  0.853457     -1.476652   
2    18.0   1.495535      1.193953        150  0.549975     -1.658700   
3    16.0   1.495535      1.059658        150  0.546432     -1.294604   
4    17.0   1.495535      1.040473        140  0.565326     -1.840747   
..    ...        ...           ...        ...       ...           ...   
392  27.0  -0.858191     -0.513513         86 -0.212865      0.016141   
393  44.0  -0.858191     -0.925991         52 -0.992237      3.293004   
394  32.0  -0.858191     -0.561476         84 -0.797394     -1.440242   
395  28.0  -0.858191     -0.705363         79 -0.407708      1.108429   
396  31.0  -0.858191     -0.714956         82 -0.295526      1.399705   

         year  origin                       name  high_mpg  
0   -1.626699       1  chevrolet chevelle malibu         0  
1   -1.626699       1          buick skylark 320         0  
2   -1.626699       1         plymouth satellite         0  
3   -1.626699       1              amc rebel sst         0  
4   -1.626699       1                ford torino         0  
..        ...     ...                        ...       ...  
392  1.629433       1            ford mustang gl         1  
393  1.629433       2                  vw pickup         1  
394  1.629433       1              dodge rampage         1  
395  1.629433       1                ford ranger         1  
396  1.629433       1                 chevy s-10         1  

[397 rows x 10 columns]

      mpg  cylinders  displacement horsepower    weight  acceleration  \
0    18.0   1.495535      1.088435        130  0.630274     -1.294604   
1    15.0   1.495535      1.500913        165  0.853457     -1.476652   
2    18.0   1.495535      1.193953        150  0.549975     -1.658700   
3    16.0   1.495535      1.059658        150  0.546432     -1.294604   
4    17.0   1.495535      1.040473        140  0.565326     -1.840747   
..    ...        ...           ...        ...       ...           ...   
392  27.0  -0.858191     -0.513513         86 -0.212865      0.016141   
393  44.0  -0.858191     -0.925991         52 -0.992237      3.293004   
394  32.0  -0.858191     -0.561476         84 -0.797394     -1.440242   
395  28.0  -0.858191     -0.705363         79 -0.407708      1.108429   
396  31.0  -0.858191     -0.714956         82 -0.295526      1.399705   

         year  origin                       name  high_mpg  
0   -1.626699       1  chevrolet chevelle malibu         0  
1

In [ ]:
# Importing necessary libraries
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVC
from sklearn.model_selection import cross_val_score

# Load the dataset
file_path = '/content/Auto.csv'
auto_df = pd.read_csv(file_path)

# Identify features for scaling (excluding non-numeric and target variables)
features_to_scale = ['cylinders', 'displacement', 'weight', 'acceleration', 'year']

# Scale features to mean 0 and standard deviation 1
scaler = StandardScaler()
auto_df[features_to_scale] = scaler.fit_transform(auto_df[features_to_scale])

# Create binary variable for 'mpg' above/below median
mpg_median = auto_df['mpg'].median()
auto_df['high_mpg'] = (auto_df['mpg'] > mpg_median).astype(int)

# Define feature set (excluding 'mpg' and non-numeric columns) and target variable
X = auto_df.drop(columns=['mpg', 'name', 'origin', 'high_mpg', 'horsepower'])  # Exclude mpg and non-relevant columns
y = auto_df['high_mpg']

# List of different values for the regularization parameter C
C_values = [0.01, 0.1, 1, 10, 100]
cross_val_errors_linear = {}

# Perform cross-validation for each C value with a linear kernel
for C in C_values:
    svc_linear = SVC(C=C, kernel='linear', random_state=0)
    scores_linear = cross_val_score(svc_linear, X, y, cv=5, scoring='accuracy')
    error_linear = 1 - scores_linear.mean()  # Calculate cross-validation error
    cross_val_errors_linear[C] = error_linear

# Display the cross-validation errors for each value of C using a linear kernel
cross_val_errors_linear

{0.01: 0.10585443037974684,
 0.1: 0.15338607594936704,
 1: 0.1382911392405063,
 10: 0.1382911392405063,
 100: 0.13079113924050623}

The cross-validation errors for the support vector classifier with a linear kernel and different C values are:-

*   C = 0.01: 0.1058
*   C = 0.1: 0.153

*   C = 1: 0.138
*   C = 10: 0.138

*   C = 100: 0.1307

Analysis:-

*   Lower values of C tend to have higher cross-validation errors, indicating underfitting.
*   Higher values of C slightly reduce error, but there is minimal performance gain beyond a certain point.

*   The classifier's performance does not improve significantly with very large C values, suggesting that further increasing C may lead to diminishing accuracy.





In [ ]:
from sklearn.metrics import make_scorer, f1_score

# Define custom scorer using F1 score
f1_scorer = make_scorer(f1_score)

# Dictionary to store F1 score cross-validation errors for each value of C
cross_val_f1_errors = {}

# Perform cross-validation for each C value using the F1 score
for C in C_values:
    svc_linear = SVC(C=C, kernel='linear', random_state=0)
    scores_f1 = cross_val_score(svc_linear, X, y, cv=5, scoring=f1_scorer)
    error_f1 = 1 - scores_f1.mean()  # Calculate F1-based cross-validation error
    cross_val_f1_errors[C] = error_f1

# Display the F1 score cross-validation errors for each value of C
cross_val_f1_errors

{0.01: 0.10358018663074586,
 0.1: 0.18976238105085197,
 1: 0.179119911670929,
 10: 0.1752264278930944,
 100: 0.157453843234434}

The F1-based cross validation errors for the support vector classifier with a linear kernel and various values of C are:-

*   C = 0.01: 0.1035
*   C = 0.1: 0.189

*   C = 1: 0.179
*   C = 10: 0.175

*   C = 100: 0.157

Analysis:-

*   With the F1 score as the evaluation metric, we observe a similar trend to the accuracy-based evaluation:
       

*   Lower values of C yield the lowest error, indicating relatively good performance on balanced precision and recall.
*   Higher values continue to show slight improvements, although performance gains diminish as C increases beyond a certain point.

These results suggest that C = 0.01 might still be a balanced choice, especially if maximizing the balance between false positives and false negatives is critical.

In [ ]:
# Define new parameter values for gamma in radial basis function (RBF) kernel and degree for polynomial kernel
gamma_values = [0.01, 0.1, 1]
degree_values = [2, 3, 4]

# Dictionaries to store F1 score cross-validation errors for each kernel and parameter configuration
cross_val_f1_errors_rbf = {}
cross_val_f1_errors_poly = {}

# Evaluate SVM with radial basis function (RBF) kernel using different values of C and gamma
for C in C_values:
    for gamma in gamma_values:
        svc_rbf = SVC(C=C, kernel='rbf', gamma=gamma, random_state=0)
        scores_f1_rbf = cross_val_score(svc_rbf, X, y, cv=5, scoring=f1_scorer)
        error_f1_rbf = 1 - scores_f1_rbf.mean()  # Calculate F1-based cross-validation error
        cross_val_f1_errors_rbf[(C, gamma)] = error_f1_rbf

# Evaluate SVM with polynomial kernel using different values of C and degree
for C in C_values:
    for degree in degree_values:
        svc_poly = SVC(C=C, kernel='poly', degree=degree, random_state=0)
        scores_f1_poly = cross_val_score(svc_poly, X, y, cv=5, scoring=f1_scorer)
        error_f1_poly = 1 - scores_f1_poly.mean()  # Calculate F1-based cross-validation error
        cross_val_f1_errors_poly[(C, degree)] = error_f1_poly

# Display the F1 score cross-validation errors for each configuration of RBF and polynomial kernels
cross_val_f1_errors_rbf, cross_val_f1_errors_poly

({(0.01, 0.01): 1.0,
  (0.01, 0.1): 0.24330404089581303,
  (0.01, 1): 1.0,
  (0.1, 0.01): 0.1266180082736783,
  (0.1, 0.1): 0.10473927743004052,
  (0.1, 1): 0.22738348347478488,
  (1, 0.01): 0.10555296783277213,
  (1, 0.1): 0.20689270558471606,
  (1, 1): 0.22332517978590705,
  (10, 0.01): 0.18813894665734077,
  (10, 0.1): 0.22738244397865714,
  (10, 1): 0.2593001320264281,
  (100, 0.01): 0.18768428001891413,
  (100, 0.1): 0.22459752127847654,
  (100, 1): 0.24947484633926764},
 {(0.01, 2): 0.8641998734977863,
  (0.01, 3): 0.22176171499758746,
  (0.01, 4): 0.3069628802001827,
  (0.1, 2): 0.5384020016680566,
  (0.1, 3): 0.30308335226110117,
  (0.1, 4): 0.30780828779374336,
  (1, 2): 0.5495049504950495,
  (1, 3): 0.22672605352683228,
  (1, 4): 0.5454695499270553,
  (10, 2): 0.5165956534428953,
  (10, 3): 0.17557244895346835,
  (10, 4): 0.5199441002410126,
  (100, 2): 0.5024040686823442,
  (100, 3): 0.18024186758515115,
  (100, 4): 0.5098185606159251})

The F1-based cross-validation errors for the SVM models with radial basis finction (RBF) and polynomial kernels across various values of C, gamma, and degree are as follows:-

Radial Basis Function (RBF) Kernel:-

*   Optimal Errors:- Lower F1-score errors are achieved with moderate C values and gamma values around 0.1.
*   High Error at Low Gamma:- Small gamma values with low C result in high errors suggesting underfitting.

*   Overfitting:- Higher gamma values, combined with large C values, tend to increase the error slightly, suggesting that an excessively complex boundary leads to overfitting.

Polynomial Kernel:-

*   Degree Effects:- The polynomial kernel shows variability in performance based on the degree. Lower degrees(like 2) generally achievelower F1 errors compared to higher degrees.
*   Optimal Parameters:- Moderate C values combined with degree 2 or 3 offer reasonable performance, with C = 1 and degree 2 achieving an F1-error around 0.55.

*   Complex Boundaries:- Higher degrees or larger C values result in increased error, indicating potential overfitting


Conclusion:-

The RBF kernel generally performs better with moderate C and gamma values. The polynomial kernel can also be effective, but higher degrees do not improve F1-score and may lead to overfitting.